chunks/chunks.go

```diff
@@ -57,8 +57,10 @@ type Meta struct {
 }
 
 // writeHash writes the chunk encoding and raw data into the provided hash.
-func (cm *Meta) writeHash(h hash.Hash) error {
-	if _, err := h.Write([]byte{byte(cm.Chunk.Encoding())}); err != nil {
+// 'buf' byte slice is used to avoid allocating new byte slice. Assumes that len(buf) is 0.
+func (cm *Meta) writeHash(h hash.Hash, buf []byte) error {
+	buf = append(buf[:0], byte(cm.Chunk.Encoding()))
```

Comment:
Hm.. are we trying to save here one byte of allocation per ops? (: Also not sure if it really assumes that `len(buf)` is 0 right?

Reply:
I forgot the remove the old comment (removed now), we don't assume `len(buf)` to be `0` now.

> Hm.. are we trying to save here one byte of allocation per ops

Nope. Slice header = 2 bytes (?), 1 byte for the data. And this is called per series per chunk. So it would be millions of allocs at a large scale. This is the benchmark difference just for the `writeHash` (the base contains optimization in `WriteChunks`).
```
benchmark                                                                                 old allocs     new allocs     delta
BenchmarkCompaction/type=normal,blocks=2,series=1000000,samplesPerSeriesPerBlock=51-8     14825915       12829175       -13.47%
```

Apparently, in my benchmark, about 60-67% of allocs savings is from `writeHash` and remaining from re-using buffer in `WriteChunks`.

Reply:
Wouldn't new slice headers be created when passing the slice around, so the only saving is the single byte? That said, I can see how this being in an inner loop could save a lot of gc.

Reply:
> Wouldn't new slice headers be created when passing the slice around

Interesting, I didn't know/think about it!

Reply:
> Nope. Slice header = 2 bytes (?), 

https://golang.org/pkg/reflect/#SliceHeader

I can see 3 integers so it's definitely not 2 bytes -> rather 12 if those as those probably are int64 (:

Reply:
Shows that I was in a hurry :sweat_smile:. I considered 1 byte per variable without much thought!

The savings in allocs with this is good anyway, so WDYT?

Reply:
Yup :+1: